File: be-service/app/models/market_entry.py
from pydantic import BaseModel, model_validator
from typing import Optional


class MarketEntryRequest(BaseModel):
    product_text: Optional[str] = None
    category: Optional[str] = None

    @model_validator(mode="after")
    def require_one(self):
        if not self.product_text and not self.category:
            raise ValueError("Either product_text or category is required")
        return self


class PriceRange(BaseModel):
    p25: float
    median: float
    p75: float


class TierInfo(BaseModel):
    tier: str
    shop_count: int
    gap_share: float
    is_gap: bool
    price_range: PriceRange


class MarketEntryResult(BaseModel):
    category: str
    confidence: float
    n_competitors: int
    recommended_tier: Optional[str]
    seller_share: str
    entry_price_range: PriceRange
    all_tiers: list[TierInfo]


class CategorySuggestion(BaseModel):
    category: str
    score: float


class LowConfidenceResult(BaseModel):
    status: str
    message: str
    top_categories: list[CategorySuggestion]

File: fe-service/src/lib/marketEntry.ts
import api from "./api";

export interface PriceRange {
  p25: number;
  median: number;
  p75: number;
}

export interface TierInfo {
  tier: string;
  shop_count: number;
  gap_share: number;
  is_gap: boolean;
  price_range: PriceRange;
}

export interface MarketEntryResult {
  category: string;
  confidence: number;
  n_competitors: number;
  recommended_tier: string | null;
  seller_share: string;
  entry_price_range: PriceRange;
  all_tiers: TierInfo[];
  trend_signal?: number;
  trend_direction?: "rising" | "flat";
}

export interface CategorySuggestion {
  category: string;
  score: number;
}

export interface LowConfidenceResult {
  status: "low_confidence";
  message: string;
  top_categories: CategorySuggestion[];
}

export type MarketEntryResponse = MarketEntryResult | LowConfidenceResult;

export const CATEGORIES = [
  "Fashion",
  "Electronics",
  "Sports",
  "Automotive",
  "Beauty",
  "Home & Living",
  "Books",
  "Food & Beverage",
  "Toys",
  "Pets",
  "Tools",
] as const;

export type Category = (typeof CATEGORIES)[number];

export function getGapShare(result: MarketEntryResult): number {
  return result.all_tiers.find((t) => t.is_gap)?.gap_share ?? 1;
}

export async function analyzeMarketEntry(
  productText: string
): Promise<MarketEntryResponse> {
  const res = await api.post("/api/v1/market-entry/analyze", {
    product_text: productText,
  });
  return res.data.data;
}

export async function analyzeMultipleCategories(
  categories: string[]
): Promise<MarketEntryResult[]> {
  const results = await Promise.all(categories.map(getMarketEntryByCategory));
  return results.sort((a, b) => getGapShare(a) - getGapShare(b));
}

export async function getMarketEntryByCategory(
  category: string
): Promise<MarketEntryResult> {
  const res = await api.post("/api/v1/market-entry/analyze", { category });
  return res.data.data;
}

export function formatRupiah(amount: number): string {
  if (amount >= 1_000_000) return `Rp ${(amount / 1_000_000).toFixed(1)}jt`;
  if (amount >= 1_000) return `Rp ${(amount / 1_000).toFixed(0)}k`;
  return `Rp ${amount}`;
}

export function formatPriceRange(range: PriceRange): string {
  return `${formatRupiah(range.p25)} – ${formatRupiah(range.p75)}`;
}
